src/parse-steps.js

```diff
@@ -0,0 +1,76 @@
+import { getSlides, getCodes } from "./differ";
+import { parseFocus as newParseFocus } from "./focus-parser";
+
+export function parseSteps(rawSteps, lang) {
+  const codes = getCodes(rawSteps);
+
+  const stepsLines = getSlides(codes.reverse(), lang).reverse();
+  const steps = rawSteps.map((step, i) => {
+    const lines = stepsLines[i];
+    return parseStep(step, lines);
+  });
+
+  steps.forEach(step => {
+    const { lines, focusMap } = step;
+    lines.forEach((line, index) => {
+      line.focus = focusMap.has(index);
+      const columnFocus = focusMap.get(index);
+      line.focusPerToken = Array.isArray(columnFocus);
+      if (line.focusPerToken) {
+        // this mutates the tokens array in order to change it to the same line in other steps
+        splitTokensToColumns(line.tokens);
+        line.tokens = setTokenFocus(line.tokens, columnFocus);
+      }
+    });
+  });
+
+  return steps;
+}
+
+function parseStep(step, lines) {
+  const { focus, ...rest } = step;
+  let focusMap = newParseFocus(focus);
+
+  if (!focusMap) {
+    // default focus
+    const indexes = lines
+      .filter(line => line.isNew)
+      .map((line, index) => index);
```

Comment:
I think this is incorrect, the index here is after filtering so it no longer corresponds to the line number.

Reply:
You are right, thanks